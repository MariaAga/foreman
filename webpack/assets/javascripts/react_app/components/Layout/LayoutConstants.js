export const LAYOUT_INITIALIZE = 'LAYOUT_INITIALIZE';
export const LAYOUT_SHOW_LOADING = 'LAYOUT_SHOW_LOADING';
export const LAYOUT_HIDE_LOADING = 'LAYOUT_HIDE_LOADING';
export const LAYOUT_CHANGE_IS_NAV_OPEN = 'LAYOUT_CHANGE_IS_NAV_OPEN';
