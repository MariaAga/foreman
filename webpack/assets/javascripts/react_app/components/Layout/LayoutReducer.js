import Immutable from 'seamless-immutable';

import {
  LAYOUT_INITIALIZE,
  LAYOUT_SHOW_LOADING,
  LAYOUT_HIDE_LOADING,
  LAYOUT_CHANGE_IS_NAV_OPEN,
} from './LayoutConstants';

const initialState = Immutable({
  items: [],
  isLoading: false,
  isNavOpen: true,
});

export default (state = initialState, action) => {
  const { payload, type } = action;

  switch (type) {
    case LAYOUT_INITIALIZE:
      return state
        .set('items', payload.items)
        .set('isNavOpen', payload.isNavOpen)
        .set('currentOrganization', payload.organization)
        .set('currentLocation', payload.location);

    case LAYOUT_SHOW_LOADING:
      return state.set('isLoading', true);

    case LAYOUT_HIDE_LOADING:
      return state.set('isLoading', false);

    case LAYOUT_CHANGE_IS_NAV_OPEN:
      return state.set('isNavOpen', payload.isNavOpen);

    default:
      return state;
  }
};
