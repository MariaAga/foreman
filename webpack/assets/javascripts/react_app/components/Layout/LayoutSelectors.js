import { get, snakeCase } from 'lodash';
import { noop } from '../../common/helpers';
import { deprecate } from '../../common/DeprecationService';

export const selectLayout = state => state.layout;

export const selectMenuItems = state => selectLayout(state).items;
export const selectIsLoading = state => selectLayout(state).isLoading;
export const selectIsNavOpen = state => selectLayout(state).isNavOpen;
export const selectCurrentLocation = state => {
  deprecate('selectCurrentLocation', 'useForemanLocation hook', 2.5);
  return get(selectLayout(state), 'currentLocation');
};
export const selectCurrentOrganization = state => {
  deprecate('selectCurrentOrganization', 'useForemanOrganization hook', 2.5);
  return get(selectLayout(state), 'currentOrganization');
};

export const patternflyMenuItemsSelector = state => {
  const items = selectMenuItems(state);
  return items.map(item => {
    const childrenArray = item.children
      .filter(child => child.name)
      .map(child => childToMenuItem(child));

    return {
      title: item.name,
      iconClass: item.icon,
      subItems: childrenArray,
      className: item.className,
    };
  });
};

const childToMenuItem = child => ({
  id: `menu_item_${snakeCase(child.name)}`,
  title: child.name,
  isDivider: child.type === 'divider',
  href: child.url || '#',
  preventHref: true,
  onClick: child.onClick || noop,
  isActive: child.isActive,
});
