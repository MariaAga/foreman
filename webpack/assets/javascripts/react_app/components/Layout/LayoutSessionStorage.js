import { getValue, setValue } from '../../common/SessionStorage';

export const getIsNavbarOpen = () => !!getValue('navOpen');

export const setIsNavbarOpen = value => setValue('navOpen', value);
