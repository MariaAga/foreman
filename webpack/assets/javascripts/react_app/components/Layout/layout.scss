@import '~@theforeman/vendor/scss/variables';
@import '../../common/colors.scss';
@import '../../common/variables.scss';

.pf-c-page {
  background-color: rgba(0, 0, 0, 0);
}

#react-app-root {
  .pf-c-page__main section {
    padding: 0 40px;
  }
}

/*
  patternfly-sass hides the vertical-nav on smaller screens, see: https://github.com/patternfly/patternfly-sass/blob/5e7cce3445d5b1af1e16500695b9f33edc6a4f6c/assets/javascripts/patternfly-functions.js#L260
  the react-container needs to be shifted when it happens.
*/
@media only screen and (max-width: $pf-global--breakpoint--md) {
  #react-app-root .react-container,
  #rails-app-content {
    margin-left: 0;
  }
}

#vertical-spinner {
  margin-top: 15px;
  margin-left: 5px;
}

.collapsed-nav {
  #react-app-root .react-container,
  #rails-app-content {
    margin-left: 75px;
  }

  @media only screen and (max-width: $pf-global--breakpoint--md) {
    #react-app-root .react-container,
    #rails-app-content {
      margin-left: 0;
    }
  }
}

#page-sidebar {
  .pf-c-nav__item.hidden-nav-lg {
    display: none;

    @media (max-width: $pf-global--breakpoint--lg) {
      display: block;
    }
  }
}
