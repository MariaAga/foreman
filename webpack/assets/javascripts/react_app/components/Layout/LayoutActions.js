import {
  LAYOUT_INITIALIZE,
  LAYOUT_SHOW_LOADING,
  LAYOUT_HIDE_LOADING,
  LAYOUT_CHANGE_IS_NAV_OPEN,
} from './LayoutConstants';

import { setIsNavbarOpen } from './LayoutSessionStorage';

export const initializeLayout = ({
  items,
  isNavOpen,
  organization,
  location,
}) => ({
  type: LAYOUT_INITIALIZE,
  payload: {
    items,
    isNavOpen,
    organization,
    location,
  },
});

export const showLoading = () => ({
  type: LAYOUT_SHOW_LOADING,
});

export const hideLoading = () => ({
  type: LAYOUT_HIDE_LOADING,
});

export const changeIsNavOpen = isNavOpen => {
  setIsNavbarOpen(isNavOpen);
  return {
    type: LAYOUT_CHANGE_IS_NAV_OPEN,
    payload: {
      isNavOpen,
    },
  };
};
