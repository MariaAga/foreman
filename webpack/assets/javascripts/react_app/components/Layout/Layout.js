import React from 'react';
import {
  Brand,
  Page,
  PageSection,
  PageSectionVariants,
  PageSidebar,
  SkipToContent,
  PageHeader,
} from '@patternfly/react-core';

import { translate as __ } from '../../common/I18n';
import VerticalNav from './components/VerticalNav';
import HeaderToolbar from './components/Toolbar/HeaderToolbar';

import { layoutPropTypes, layoutDefaultProps } from './LayoutHelper';
import './layout.scss';

const Layout = ({
  items,
  data,
  isLoading,
  isNavOpen,
  navigate,
  children,
  changeIsNavOpen,
}) => {
  const onNavToggle = () => {
    isNavOpen ? changeIsNavOpen(false) : changeIsNavOpen(true);
  };
  const header = (
    <PageHeader
      logo={
        <React.Fragment>
          <Brand src={data.logo} alt={data.brand} href={data.root} />
          <span className="navbar-brand-txt">
            <span>{data.brand}</span>
          </span>
        </React.Fragment>
      }
      logoProps={{ href: data.root }}
      showNavToggle
      isManagedSidebar
      onNavToggle={onNavToggle}
      headerTools={<HeaderToolbar {...data} isLoading={isLoading} />}
    />
  );

  const sidebar = (
    <PageSidebar
      isNavOpen={isNavOpen}
      nav={<VerticalNav items={items} navigate={navigate} />}
    />
  );
  const pageId = 'main';
  const pageSkipToContent = (
    <SkipToContent href={`#${pageId}`}>{__('Skip to content')}</SkipToContent>
  );
  return (
    <Page
      header={header}
      sidebar={sidebar}
      skipToContent={pageSkipToContent}
      mainContainerId={pageId}
    >
      <PageSection
        className="react-container nav-pf-persistent-secondary"
        variant={PageSectionVariants.light}
      >
        {children}
      </PageSection>
    </Page>
  );
};
Layout.propTypes = layoutPropTypes;

Layout.defaultProps = layoutDefaultProps;

export default Layout;
