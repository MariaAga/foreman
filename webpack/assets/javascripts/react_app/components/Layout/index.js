import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';

import {
  initializeLayout,
  collapseLayoutMenus,
  expandLayoutMenus,
  changeIsNavOpen,
} from './LayoutActions';
import reducer from './LayoutReducer';
import {
  patternflyMenuItemsSelector,
  selectIsLoading,
  selectIsNavOpen,
} from './LayoutSelectors';
import { combineMenuItems } from './LayoutHelper';

import { getIsNavbarOpen } from './LayoutSessionStorage';

import Layout from './Layout';

const ConnectedLayout = ({ children, data }) => {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(
      initializeLayout({
        items: combineMenuItems(data),
        isNavOpen: getIsNavbarOpen(),
        organization: data.orgs.current_org,
        location: data.locations.current_location,
      })
    );
  }, [data, dispatch]);

  const isNavOpen = useSelector(state => selectIsNavOpen(state));
  useEffect(() => {
    // toggles a class in the body tag, so that the main #rails-app-content container can have the appropriate width
    if (isNavOpen) {
      document.body.classList.add('pf-m-expanded');
    } else {
      document.body.classList.remove('pf-m-expanded');
    }
  }, [isNavOpen]);

  const { push: navigate } = useHistory();
  const items = useSelector(state => patternflyMenuItemsSelector(state));
  const isLoading = useSelector(state => selectIsLoading(state));
  return (
    <Layout
      data={data}
      navigate={navigate}
      items={items}
      isLoading={isLoading}
      isNavOpen={isNavOpen}
      collapseLayoutMenus={() => dispatch(collapseLayoutMenus())}
      expandLayoutMenus={() => dispatch(expandLayoutMenus())}
      changeIsNavOpen={value => dispatch(changeIsNavOpen(value))}
    >
      {children}
    </Layout>
  );
};

// export prop-types
export const { propTypes, defaultProps } = Layout;

ConnectedLayout.propTypes = {
  children: propTypes.children,
  data: propTypes.data,
};

ConnectedLayout.defaultProps = {
  children: defaultProps.children,
  data: defaultProps.data,
};

// export reducers
export const reducers = { layout: reducer };

// export connected component
export default ConnectedLayout;
