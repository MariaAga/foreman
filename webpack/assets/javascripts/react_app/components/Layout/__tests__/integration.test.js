import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';

import { IntegrationTestHelper } from '@theforeman/test';

import { hasTaxonomiesMock } from '../Layout.fixtures';
import Layout, { reducers } from '../index';

jest.mock('../../notifications', () => 'span');

describe('Layout integration test', () => {
  it('should flow', async () => {
    const integrationTestHelper = new IntegrationTestHelper(reducers);

    const component = integrationTestHelper.mount(
      <Router>
        <Layout {...hasTaxonomiesMock} />
      </Router>
    );
    await IntegrationTestHelper.flushAllPromises();
    component.update();

    integrationTestHelper.takeStoreSnapshot('initial state');
    const hostsMenuItem = component.find('.secondary-nav-item-pf > a');
    const locationToggle = component.find(
      '#location-dropdown .pf-c-context-selector__toggle'
    );
    const organizationToggle = component.find(
      '#organization-dropdown .pf-c-context-selector__toggle'
    );
    locationToggle.simulate('click');
    component.update();
    const yamlLocation = component.find('#select_taxonomy_yaml');
    yamlLocation.at(0).simulate('click');
    integrationTestHelper.takeStoreAndLastActionSnapshot(
      'Location "yaml" clicked'
    );
    expect(
      component
        .find('#location-dropdown .pf-c-context-selector__toggle-text')
        .text()
    ).toBe('yaml');

    organizationToggle.simulate('click');
    component.update();
    const org2Organization = component.find('#select_taxonomy_org2');
    org2Organization.at(0).simulate('click');
    integrationTestHelper.takeStoreAndLastActionSnapshot('Org "org2" clicked');
    expect(
      component
        .find('#organization-dropdown .pf-c-context-selector__toggle-text')
        .text()
    ).toBe('org2');

    await IntegrationTestHelper.flushAllPromises();
    component.update();

    hostsMenuItem.at(1).simulate('click');
    integrationTestHelper.takeStoreAndLastActionSnapshot(
      'Changed ActiveMenu to Hosts'
    );
    expect(component.find('.secondary-nav-item-pf .active > a').text()).toBe(
      'Hosts'
    );
  });
});
