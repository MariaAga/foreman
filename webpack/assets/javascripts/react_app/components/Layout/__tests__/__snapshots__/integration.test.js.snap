// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout integration test should flow: initial state 1`] = `
Object {
  "layout": Object {
    "currentLocation": "london",
    "currentOrganization": "org1",
    "isLoading": false,
    "isNavOpen": false,
    "items": Array [
      Object {
        "children": Array [
          Object {
            "exact": true,
            "name": "Dashboard",
            "type": "item",
            "url": "/",
          },
          Object {
            "name": "Facts",
            "type": "item",
            "url": "/fact_values",
          },
        ],
        "className": "",
        "icon": "fa fa-tachometer",
        "name": "Monitor",
        "type": "sub_menu",
      },
      Object {
        "children": Array [
          Object {
            "name": "All Hosts",
            "type": "item",
            "url": "/hosts/new",
          },
        ],
        "className": "",
        "icon": "fa fa-server",
        "name": "Hosts",
        "type": "sub_menu",
      },
      Object {
        "children": Array [
          Object {
            "name": undefined,
            "type": "item",
            "url": "/nameless",
          },
          Object {
            "name": undefined,
            "type": "divider",
          },
        ],
        "className": "",
        "icon": "pficon pficon-unplugged",
        "name": "Empty",
        "type": "sub_menu",
      },
      Object {
        "children": Array [
          Object {
            "isActive": false,
            "name": "Any Organization",
            "onClick": [Function],
          },
          Object {
            "isActive": true,
            "name": "org1",
            "onClick": [Function],
            "type": undefined,
          },
          Object {
            "isActive": false,
            "name": "org2",
            "onClick": [Function],
            "type": undefined,
          },
        ],
        "className": "organization-menu hidden-nav-lg",
        "icon": "fa fa-building",
        "name": "Organizations",
        "type": "sub_menu",
      },
      Object {
        "children": Array [
          Object {
            "isActive": false,
            "name": "Any Location",
            "onClick": [Function],
          },
          Object {
            "isActive": false,
            "name": "yaml",
            "onClick": [Function],
            "type": undefined,
          },
          Object {
            "isActive": true,
            "name": "london",
            "onClick": [Function],
            "type": undefined,
          },
          Object {
            "isActive": false,
            "name": "norway",
            "onClick": [Function],
            "type": undefined,
          },
        ],
        "className": "location-menu hidden-nav-lg",
        "icon": "fa fa-globe",
        "name": "Locations",
        "type": "sub_menu",
      },
    ],
  },
}
`;
